controllers/apim/virtualservice_controller.go

```diff
@@ -0,0 +1,216 @@
+			operation := toRule.Operation
+			if len(operation.Paths) == 0 && len(operation.Methods) == 0 && idx > 0 {
+				continue
+			}
+			authToRules = append(authToRules, toRule)
+		}
+	}
+
+	authPolicy.Spec = securityv1beta1.AuthorizationPolicy{
+		Rules: []*securityv1beta1.Rule{{
+			To: authToRules,
+		}},
+		Action: securityv1beta1.AuthorizationPolicy_CUSTOM,
+		ActionDetail: &securityv1beta1.AuthorizationPolicy_Provider{
+			Provider: &securityv1beta1.AuthorizationPolicy_ExtensionProvider{
+				Name: providerName,
+			},
+		},
+	}
+
+	if err := controllerutil.SetOwnerReference(vs, &authPolicy, r.Client().Scheme()); err != nil {
+		logger.Error(err, "failed to add owner ref to AuthorizationPolicy resource")
+		return err
+	}
+	err := r.ReconcileResource(ctx, &istiosecurityv1beta1.AuthorizationPolicy{}, &authPolicy, alwaysUpdateAuthPolicy)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		logger.Error(err, "ReconcileResource failed to create/update AuthorizationPolicy resource")
+		return err
+	}
+
+	logger.Info("successfully created/updated AuthorizationPolicy resources")
+	return nil
+}
+
+func alwaysUpdateAuthPolicy(existingObj, desiredObj client.Object) (bool, error) {
+	existing, ok := existingObj.(*istiosecurityv1beta1.AuthorizationPolicy)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *istiosecurityv1beta1.AuthorizationPolicy", existingObj)
+	}
+	desired, ok := desiredObj.(*istiosecurityv1beta1.AuthorizationPolicy)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *istiosecurityv1beta1.AuthorizationPolicy", desiredObj)
+	}
+
+	existing.Spec = desired.Spec
+	return true, nil
+}
+
+func normalizeStringMatch(sm *networkingv1alpha3.StringMatch) string {
+	if prefix := sm.GetPrefix(); prefix != "" {
+		return prefix + "*"
+	}
+	if exact := sm.GetExact(); exact != "" {
+		return exact
+	}
+	// Regex string match is not supported because authpolicy doesn't as well.
+	return ""
+}
+
+// VirtualServiceFilter allows generation of only relevant reconciliation request.
+type VirtualServiceFilter struct {
+	K8sClient client.Client
+	Logger    logr.Logger
+}
+
+// Map contains filtering logic for virtualservice.
+func (m *VirtualServiceFilter) Map(obj client.Object) []reconcile.Request {
+	virtualServiceAnnotations := obj.GetAnnotations()
+	_, present := virtualServiceAnnotations[KuadrantAuthProviderAnnotation]
+	if !present {
+		authObjKey := types.NamespacedName{
+			Name:      getAuthPolicyName(obj.GetName()),
+			Namespace: obj.GetNamespace(),
+		}
+
+		var authPolicy client.Object
+		if err := m.K8sClient.Get(context.Background(), authObjKey, authPolicy); err != nil {
+			// no annotation but authpolicy exist means annotation was removed
+			if !apierrors.IsNotFound(err) {
+				m.Logger.Error(err, "failed to check AuthorizationPolicy existence")
+			}
+			return nil // this virtualservice is not protected
+		}
+
+		// AuthorizationPolicy exists
+		if err := m.K8sClient.Delete(context.Background(), authPolicy); err != nil {
```

Comment:
this feels weird in a map function should we not just return the reconcile request and handle this in the reconcile logic?

Reply:
Creating a request for it means creating for all the events. Shouldn't we only care about virtualservices with annotation in the reconcile logic?

Reply:
I can move it if you want?

Reply:
I agree with @maleck13 . It looks like this is changing the state of the cluster while filtering the resources.

Reply:
Filter is now done in the reconcile logic along with the orphan check.

Reply:
I remembered that I added the ownerRef to the AuthPolicy so, in theory, orphan authpolicy shouldn't exist at all. I'll just remove that check :thinking:

Reply:
Done.

Reply:
Sorry, my bad, I added orphan check because let's say the user removes the annotation, and there is no automatic removal of AuthPolicy. I re-added the check.